File: ainpc-core-plugin/src/main/java/ro/ainpc/world/patch/PatchCandidate.java
package ro.ainpc.world.patch;

import java.util.List;

public record PatchCandidate(
    String candidateId,
    PatchGapType gapType,
    PatchType patchType,
    String targetRegionId,
    String targetPlaceId,
    int priority,
    int cost,
    int risk,
    List<String> requiredCapabilities,
    String reason
) {
    public PatchCandidate {
        candidateId = valueOrEmpty(candidateId);
        targetRegionId = valueOrEmpty(targetRegionId);
        targetPlaceId = valueOrEmpty(targetPlaceId);
        priority = Math.max(0, priority);
        cost = Math.max(0, cost);
        risk = Math.max(0, risk);
        requiredCapabilities = List.copyOf(requiredCapabilities != null ? requiredCapabilities : List.of());
        reason = valueOrEmpty(reason);
    }

    private static String valueOrEmpty(String value) {
        return value == null ? "" : value.trim();
    }
}

File: ainpc-core-plugin/bin/main/ro/ainpc/world/patch/PatchCandidate.kt
package ro.ainpc.world.patch

class PatchCandidate(
    candidateId: String?,
    private val gapType: PatchGapType,
    private val patchType: PatchType,
    targetRegionId: String?,
    targetPlaceId: String?,
    priority: Int,
    cost: Int,
    risk: Int,
    requiredCapabilities: List<String>?,
    reason: String?
) {
    private val candidateId: String = valueOrEmpty(candidateId)
    private val targetRegionId: String = valueOrEmpty(targetRegionId)
    private val targetPlaceId: String = valueOrEmpty(targetPlaceId)
    private val priority: Int = maxOf(0, priority)
    private val cost: Int = maxOf(0, cost)
    private val risk: Int = maxOf(0, risk)
    private val requiredCapabilities: List<String> = (requiredCapabilities ?: emptyList()).toList()
    private val reason: String = valueOrEmpty(reason)

    fun candidateId(): String = candidateId
    fun gapType(): PatchGapType = gapType
    fun patchType(): PatchType = patchType
    fun targetRegionId(): String = targetRegionId
    fun targetPlaceId(): String = targetPlaceId
    fun priority(): Int = priority
    fun cost(): Int = cost
    fun risk(): Int = risk
    fun requiredCapabilities(): List<String> = requiredCapabilities
    fun reason(): String = reason

    companion object {
        private fun valueOrEmpty(value: String?): String = value?.trim().orEmpty()
    }
}
